src/NHibernate/Impl/AsyncEnumerableImpl.cs

```diff
@@ -0,0 +1,258 @@
+				_initializeAsync,
+				_readOnly,
+				_types,
+				_columnNames,
+				_selection,
+				_resultTransformer,
+				_returnAliases,
+				_session,
+				cancellationToken);
+		}
+
+		private sealed class AsyncEnumeratorImpl : IAsyncEnumerator<T>
+		{
+			private static readonly INHibernateLogger log = NHibernateLogger.For(typeof(AsyncEnumeratorImpl));
+
+			private readonly InitializeAsync _initializeAsync;
+			private readonly bool _readOnly;
+			private readonly IType[] _types;
+			private string[][] _columnNames;
+			private readonly RowSelection _selection;
+			private readonly IResultTransformer _resultTransformer;
+			private readonly string[] _returnAliases;
+			private readonly IEventSource _session;
+			private readonly CancellationToken _cancellationToken;
+			private readonly bool _single;
+
+			private DbDataReader _reader;
+			private DbCommand _command;
+			private bool _hasNext;
+			private bool _startedReading; // True if at least one MoveNext call was made.
+			// when we start enumerating through the DataReader we are positioned
+			// before the first record we need
+			private int _currentRow = -1;
+			private bool _isAlreadyDisposed;
+
+			public AsyncEnumeratorImpl(
+				InitializeAsync initializeAsync,
+				bool readOnly,
+				IType[] types,
+				string[][] columnNames,
+				RowSelection selection,
+				IResultTransformer resultTransformer,
+				string[] returnAliases,
+				IEventSource session,
+				CancellationToken cancellationToken)
+			{
+				_initializeAsync = initializeAsync;
+				_readOnly = readOnly;
+				_types = types;
+				_single = types.Length == 1;
+				_columnNames = columnNames;
+				_selection = selection;
+				_resultTransformer = resultTransformer;
+				_returnAliases = returnAliases;
+				_session = session;
+				_cancellationToken = cancellationToken;
+			}
+
+			public T Current { get; private set; }
+
+			public async ValueTask<bool> MoveNextAsync()
+			{
+				if (_reader == null)
+				{
+					await InitializeAsync().ConfigureAwait(false);
+				}
+
+				try
+				{
+					_hasNext = await _reader.ReadAsync(_cancellationToken).ConfigureAwait(false);
+					_startedReading = true;
+					_currentRow++;
+				}
+				catch (DbException e)
+				{
+					throw ADOExceptionHelper.Convert(_session.Factory.SQLExceptionConverter, e, "Error executing Enumerable() query",
+													   new SqlString(_command.CommandText));
+				}
+
+				if (_selection != null && _selection.MaxRows != RowSelection.NoValue)
+				{
+					_hasNext = _hasNext && (_currentRow < _selection.MaxRows);
+				}
+
+				bool sessionDefaultReadOnlyOrig = _session.DefaultReadOnly;
+				_session.DefaultReadOnly = _readOnly;
+				bool isCurrentValid;
+				try
+				{
+					isCurrentValid = await PostMoveNextAsync().ConfigureAwait(false);
+				}
+				finally
+				{
+					_session.DefaultReadOnly = sessionDefaultReadOnlyOrig;
+				}
+
+				// In order to simulate what SafetyEnumerable does, we have to skip elements of different type
```

Comment:
I actually quite dislike this safe enumerable part. It mostly means some user code error which is silently ignored. May be we shouldn't do it in async part?

Reply:
Same

Reply:
Now an exception will be thrown instead.